### ISSUE TITLE
Is there any way to customize the color scale range?

### ISSUE THREAD

**Issue by David-Gil** [Reading-eScience-Centre/ncwms#21]
Hi!

Is it possible to have more control over the color scale range? I'd like to determine not only the min and max values but also the intermediate ranges. For example, if the min value is 10 and the max value is 100, I'd like to be able to determine the color for the values from 10 to 19, then another color for the values from 20 to 29 and so forth.

If it's not possible (I suspect it isn't) is there a way to simulate it? Would it be possible to develop it (I'm willing to do it if necessary)?

Thank you in advance!!

David Gil

**Comment by guygriffiths** [comment]
Yes, that's entirely possible to do, you just need to create a custom style.  By default custom styles live in ~/.ncWMS2/.styles/ although different/additional directories can be defined if you prefer.

Custom styles follow the SLD spec with several custom symbolizers.  There are a few keywords which you can add which will be replaced by URL parameters.  This is how every style in ncWMS is defined, so if you want to see a variety of examples have a look here: https://github.com/Reading-eScience-Centre/edal-java/tree/master/graphics/src/main/resources/styles

What you're after sounds like the `Categorize` symbolizer we have.  The documentation for it is here: https://reading-escience-centre.gitbooks.io/edal-user-guide/content/appendices/sld_spec/xml_description/functions.html

But for standard gridded data, the style you want is probably similar to this:
```
<?xml version="1.0" encoding="ISO-8859-1"?>
<StyledLayerDescriptor version="1.1.0" xsi:schemaLocation="http://www.opengis.net/sld
StyledLayerDescriptor.xsd" xmlns="http://www.opengis.net/sld" xmlns:ogc="http://www.opengis.net/ogc" xmlns:se="http://www.opengis.net/se" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:resc="http://www.resc.reading.ac.uk">
    <NamedLayer>
        <se:Name>$layerName</se:Name>
        <UserStyle>
            <se:CoverageStyle>
                <se:Rule>
                    <se:RasterSymbolizer>
                        <se:Opacity>$opacity</se:Opacity>
                        <se:ColorMap>
                            <se:Categorize fallbackValue="$bgColor">
                                <se:LookupValue>Rasterdata</se:LookupValue>
                                <se:Value>transparent</se:Value>
                                <se:Threshold>10</se:Threshold>
                                <se:Value>blue</se:Value>
                                <se:Threshold>20</se:Threshold>
                                <se:Value>red</se:Value>
                                <se:Threshold>30</se:Threshold>
                                <se:Value>#00FF00</se:Value>
                                <se:Threshold>100</se:Threshold>
                                <se:Value>#000000</se:Value>
                            </se:Categorize>
                        </se:ColorMap>
                    </se:RasterSymbolizer>
                </se:Rule>
            </se:CoverageStyle>
        </UserStyle>
    </NamedLayer>
</StyledLayerDescriptor>
```

Just stick that in an XML file with a suitable name (the filename will be used as the style name in WMS, minus the `.xml` part) in your ~/.ncWMS2/.styles/ directory, restart, and it should be available through Godiva3.  Let me know if you have any troubles with it.

Colours can be any named colour in [this list](https://github.com/Reading-eScience-Centre/edal-java/blob/master/graphics/src/main/resources/colors.csv), or of the format #RRGGBB, or #AARRGGBB

**Further info:**
More about defining custom styles can be found here: https://reading-escience-centre.gitbooks.io/ncwms-user-guide/content/06-development.html#styles

And information about configuring different style directories is here: https://reading-escience-centre.gitbooks.io/ncwms-user-guide/content/02-installation.html#servlet

The SLD extensions we define are documented here: https://reading-escience-centre.gitbooks.io/edal-user-guide/content/appendices/sld_spec/introduction.html

**Comment by David-Gil** [comment]
Thank you so much for all this info!! There's much to learn here :-D I'll tell you something if I don't get it to work.
